hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/AbstractGauge.java

```diff
@@ -34,16 +38,95 @@ public String getName() {
         return name;
     }
 
-    void clearProbeInstance() {
-        probeInstance = null;
+    public abstract void onProbeInstanceSet(ProbeInstance probeInstance);
+
+    abstract MetricValueCatcher getCatcherOrNull();
+
+    final void onCollectionCompleted(long collectionId) {
+        if (lastCollectionId != collectionId) {
+            MetricValueCatcher catcher = getCatcherOrNull();
+            if (catcher != null && catcher instanceof AbstractMetricValueCatcher) {
+                ((AbstractMetricValueCatcher) catcher).clearCachedValue();
+                ((AbstractMetricValueCatcher) catcher).clearCachedMetricSourceRef();
+            }
+        }
+    }
+
+    abstract class AbstractMetricValueCatcher implements MetricValueCatcher {
+        private final AtomicReference<DynamicMetricSourceReference> dynamicMetricRef = new AtomicReference<>();
+
+        @Override
+        public final void catchMetricValue(long collectionId, Object source, ProbeFunction function) {
+            lastCollectionId = collectionId;
+            if (function == null || source == null) {
+                clearCachedValue();
+            } else {
+                DynamicMetricSourceReference cachedMetricRef = dynamicMetricRef.get();
+                if (cachedMetricRef == null || source != cachedMetricRef.source() || function != cachedMetricRef.function()) {
+                    dynamicMetricRef.set(new DynamicMetricSourceReference(source, function));
+                    clearCachedValue();
+                }
+            }
+        }
+
+        final <T> T readBase(MetricValueExtractorFunction<T> extractorFn, MetricCachedValueReaderFunction<T> readCachedFn) {
+            // called from LongGaugeImpl#read() and DoubleGaugeImpl#read()
+            // it does boxing and unboxing through the passed function
+            // parameters as the trade-off of the mostly shared code base
+
+            // check if we have a usable cached dynamic metric reference
+            // if so, read the value from it
+            DynamicMetricSourceReference cachedMetricRef = dynamicMetricRef.get();
+            if (cachedMetricRef != null) {
+                // need to make local copies since source and function
+                // is accessed through weak references
+                Object source = cachedMetricRef.source();
+                ProbeFunction function = cachedMetricRef.function();
+
+                if (source != null && function != null) {
+                    return extractorFn.extractValue(name, source, function, metricsRegistry);
+                } else {
+                    clearCachedMetricSourceRef();
+                }
+            }
+
+            // otherwise use the the cached value
+            return readCachedFn.readCachedValue();
```

Comment:
If I'm reading it correctly, we'll have autoboxing here. Thus some litter per each read. Can we avoid it by introducing specialized primitive classes/methods? Or I'm mistaken and it's not the case?

Reply:
Right, we autobox here. I'll think about it if we can avoid autoboxing here while keeping the logic shared. I don't worry too much about it since we have only a few gauges in the production code. Autoboxing here should be unnoticeable. Anyways, added to my improvement points list :)